lib/donation_system/salesforce/donation_validator.rb

```diff
@@ -24,19 +25,49 @@ def execute
 
       def fields
         return unless errors.empty?
+        required_data
+          .merge(card_data)
+          .merge(extra_fields)
+      end
+
+      def required_data
         {
           Amount: data.payment_data.amount.to_s,
-          CloseDate: '2017-09-11',
+          CloseDate: format_time(data.payment_data.created),
           Name: 'Online donation',
           StageName: 'Received',
           AccountId: supporter[:AccountId]
         }
       end
 
+      def card_data
+        {
+          CurrencyIsoCode: data.payment_data.currency&.upcase,
+          Last_digits__c: data.payment_data.source&.last4,
+          Card_type__c: data.payment_data.source&.brand,
+          Gateway_transaction_ID__c: data.payment_data.id,
+          Receiving_Organization__c: 'Survival UK',
+          Payment_method__c: 'Card (Stripe)',
+          RecordTypeId: '01280000000Fvqi'
+        }
+      end
+
+      def extra_fields
+        {
+          IsPrivate: false,
+          Gift_Aid__c: data.request_data.giftaid,
+          Block_Gift_Aid_Reclaim__c: false,
+          Fundraising__c: false
+        }
+      end
+
       def errors
```

Comment:
This is starting to look really procedural :/

Perhaps a better approach to this would be something like this?

```
VALIDATIONS = %i[validate_data validate_request_data .. validate_account_id].freeze

def errors
  VALIDATIONS.map { |validation| send(:validation) }.compact
end
```

Reply:
I'll think about this :thinking: